packages/dropdown-react/src/useListNavigation.ts

```diff
@@ -0,0 +1,130 @@
+import { useRef, useEffect, RefObject } from "react";
+
+type Timer = NodeJS.Timeout | undefined;
+type SearchTerm = { term: string } | undefined;
+type Direction = "prev" | "next" | "first" | "last";
+
+export function useListNavigation(enableSearch: boolean): RefObject<HTMLUListElement> {
+    const listRef = useRef<HTMLUListElement>(null);
+    let search: SearchTerm;
+    let searchResetTimer: Timer;
+    if (enableSearch) search = { term: "" };
+    useEffect(() => {
+        const listElement = listRef.current;
+        if (listElement) {
+            listElement.addEventListener("keydown", (event) =>
+                handleListKeyNav(listElement, event, search, searchResetTimer),
+            );
+        }
+    }, []);
+
+    return listRef;
+}
+
+function handleListKeyNav(list: HTMLUListElement, event: KeyboardEvent, search: SearchTerm, searchResetTimer: Timer) {
+    const { key, target } = event;
+    const currentFocus = target as HTMLButtonElement;
+    switch (key) {
+        case "ArrowUp":
+        case "PageUp":
+            event.preventDefault();
+            moveFocus(list, currentFocus, "prev");
+            break;
+        case "ArrowDown":
+        case "PageDown":
+            event.preventDefault();
+            moveFocus(list, currentFocus, "next");
+            break;
+        case "Home":
+            event.preventDefault();
+            moveFocus(list, currentFocus, "first");
+            break;
+        case "End":
+            event.preventDefault();
+            moveFocus(list, currentFocus, "last");
+            break;
+        case "Enter":
+        case " ":
+            break;
+
+        default:
+            if (search !== undefined) {
+                const searchResult = findItem(list, key, search, searchResetTimer);
+                if (searchResult) {
+                    searchResult.focus();
+                }
+            }
+            break;
+    }
+}
+
+function moveFocus(list: HTMLUListElement, current: HTMLButtonElement, dir: Direction) {
+    const thisLI = current.parentElement;
+    switch (dir) {
```

Comment:
dir? how about action? or direction? dir is directory in my mind :stuck_out_tongue: